### PR TITLE
Hugetlb and mlock support

### LINKED ISSUE #4: HugeTLB and mlock support?
Hi, I'm trying to add hugetlb and mlock support for the shared memory region to further increase the performance at https://github.com/dovahcrow/shmem-ipc. I can create a PR if you want to make these changes into the main repo. 

~Note that the implementation is not finished: I am stuck at using `set_len` to increase the size of the memfd (https://github.com/dovahcrow/shmem-ipc/blob/master/src/mem.rs#L81): it always returns EINVAL, and I haven't found a solution yet.~ By reading the Linux source code, it turned out hugetlb does not need set_len!

Any suggestion is welcomed!

### PR DESCRIPTION
**Pull request by dovahcrow**
fixes #4

### CODE REVIEW

**Comment by dovahcrow** [comment]
Two notes:
1. does this repo have rustfmt config? this will make the change less messy due to fmt differences.
2. seal does not work with hughtlb for kernels below 4.15

**Comment by diwic** [comment]
> does this repo have rustfmt config? this will make the change less messy due to fmt differences.

I've never been a fan of rustfmt, I think the code looks more ugly afterwards. That said, apparently it's a huge deal for some people and I don't want to make contributors run away because of that. I could commit a rustfmt.toml file and reformat if you want, which would then have you redo your PR because of all the changes. What's your thoughts about this?

 > seal does not work with hughtlb for kernels below 4.15

Sounds like a good reason for hugetlb to be opt-in as well, if so then it would be runnable below 4.15 by default, right? Then adding a doc comment saying that the call requires kernel 4.15+ seems like the right thing to do.

**Comment by dovahcrow** [comment]
> I've never been a fan of rustfmt, I think the code looks more ugly afterwards. That said, apparently it's a huge deal for some people and I don't want to make contributors run away because of that. I could commit a rustfmt.toml file and reformat if you want, which would then have you redo your PR because of all the changes. What's your thoughts about this?

I'm happy to follow the repo's fmt standard, as long as there is one. The problem is that when I save the file everything automatically formatted but if I don't use format on save, I need to manually insert spaces etc.. If you are ok with the changes that are due to fmt, let's keep the current PR as is. I'm also fine to rebase the code if you want to add rustfmt.

> Sounds like a good reason for hugetlb to be opt-in as well, if so then it would be runnable below 4.15 by default, right? Then adding a doc comment saying that the call requires kernel 4.15+ seems like the right thing to do.

Yep can add the doc.

P.S. just find the linux doc says 4.16+, not 4.15 https://man7.org/linux/man-pages/man2/memfd_create.2.html

**Comment by diwic** [comment]
So when looking at the content of the patch my main objection is that you're changing mmap_raw to mmap_mut. I've just opened https://github.com/RazrFalcon/memmap2-rs/pull/64 which adds mlock support to mmap_raw.

**Comment by dovahcrow** [comment]
> So when looking at the content of the patch my main objection is that you're changing mmap_raw to mmap_mut. I've just opened https://github.com/RazrFalcon/memmap2-rs/pull/64 which adds mlock support to mmap_raw.

Great!

File: src/sharedring.rs [hunk 1]
```diff
@@ -29,34 +30,80 @@ fn round_to_page_size<T>(capacity: usize) -> usize {
...
+            .close_on_exec(true);
+        if tlbsize.is_some() {
+            opts = opts.hugetlb(tlbsize);
+        }
+
         let memfd = opts.create(std::any::type_name::<T>())?;
-        memfd.as_file().set_len(bytes as u64)?;
+        if tlbsize.is_none() {
+            // hugetlb does not need/allow to set_len
+            memfd.as_file().set_len(bytes as u64)?;
```
Comment (diwic) [inline src/sharedring.rs:62]: Here's something weird. If the file's length is given by the hugetlbsize, nothing is stopping us from setting a large capacity and then a smaller hugetlbsize. This looks like it at least needs to be checked for somehow, or possibly automatically calculated?

Comment (dovahcrow) [reply src/sharedring.rs:62]: hugetlbsize is the size of a special page called huge-page. Usually, a CPU provides 2MB as the option, and some CPUs provide 1GB or others. Say, if bytes is 8MB, then this memfd will be backed by 4 2MB huge-page. hugetlbsize is not the size of the memfd.

File: src/sharedring.rs [hunk 2]
```diff
@@ -29,34 +30,80 @@ fn round_to_page_size<T>(capacity: usize) -> usize {
...
+    }
 }
 
 impl Inner {
     fn new<T>(capacity: usize) -> Result<Self, Error> {
+        Self::with_opts::<T>(capacity, None)
+    }
+
+    fn with_hugetlb<T>(capacity: usize, tlbsize: HugetlbSize) -> Result<Self, Error> {
+        Self::with_opts::<T>(capacity, Some(tlbsize))
```
Comment (diwic) [inline src/sharedring.rs]: This is a little over-engineered. Inner is not a public struct so you can just have `fn new(capacity: usize, tlbsize: Option<HugetlbSize>)` and skip `with_hugetlb` and `with_opts`.

**Comment by diwic** [comment]
@dovahcrow There is now a released version of memmap2 that allows mlock on the mmapraw struct.

File: src/mem.rs [hunk 3]
```diff
@@ -103,7 +111,7 @@ mod tests {
         assert!(memfd.seals()?.contains(&mfd::FileSeal::SealWrite));
         assert_eq!(mmap.len(), 16384);
         // The memfd is now read-only, cannot create a writable one.
-        assert!(raw_memfd(&memfd).is_err());
+        assert!(mut_memfd(&memfd).is_err());
```
Comment (diwic) [inline src/mem.rs]: This looks like you forgot to run `cargo test` before submitting the patch?

**Comment by diwic** [comment]
@dovahcrow ping - did you give up?

**Comment by dovahcrow** [comment]
> @dovahcrow ping - did you give up?

Sorry was busy lately. Just pushed a fix to the test.

**Comment by diwic** [comment]
And it's in! Thanks for the contribution. Would you like a release as well?

**Comment by dovahcrow** [comment]
> And it's in! Thanks for the contribution. Would you like a release as well?

You can decide! I'm already using the git version now it should be fine.

**Comment by diwic** [comment]
@dovahcrow Ok! I decided to run rustfmt and then release 0.3.0.

**Pull request merged**